### PR TITLE
Update driver to return ebpf_program_data_t to verifier

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Update driver to return encoded program information.
Update ebpf_api to decode program information.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: netebpfext/net_ebpf_ext.c [hunk 1]
```diff
@@ -696,8 +698,9 @@ static ebpf_error_code_t
 _net_ebpf_ext_program_information_encode_xdp()
 {
     ebpf_error_code_t return_value;
-    uint8_t* buffer = NULL;
+    const uint8_t* buffer = NULL;
     unsigned long buffer_size = 0;
+#if defined(MSRPC_LIB)
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:703]: question: when is this defined vs not defined?  Is this file compiled twice?

File: tests/end_to_end/helpers.h [hunk 2]
```diff
@@ -91,15 +93,95 @@ typedef class _program_information_provider
...
+             {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY}},
+        };
+        ebpf_context_descriptor_t context_descriptor{
+            sizeof(bind_md_t), EBPF_OFFSET_OF(bind_md_t, app_id_start), EBPF_OFFSET_OF(bind_md_t, app_id_end), -1};
+        ebpf_program_type_descriptor_t program_type_descriptor{"bind", &context_descriptor};
+        ebpf_program_information_t program_information{
+            program_type_descriptor, _countof(helper_functions), helper_functions};
+        uint8_t* buffer;
+        unsigned long buffer_size;
+        REQUIRE(ebpf_program_information_encode(&program_information, &buffer, &buffer_size) == EBPF_ERROR_SUCCESS);
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:141]: does buffer need to be freed? (same question applies to line 176)

File: netebpfext/net_ebpf_ext.c (hunk 1 above)
Comment (Alan-Jowett) [reply netebpfext/net_ebpf_ext.c:703]: Dead code I guess. This was place-holder until I got it to bind to msrpc.sys. Given that seems like a no-go I will remove this code.

File: netebpfext/net_ebpf_ext.c [hunk 3]
```diff
@@ -33,10 +33,12 @@
 #include <netiodef.h>
 #include <ntddk.h>
 
+#include "ebpf_bind_program_data.h"
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:36]: I don't see this file or ebpf_xdp_program_data.h in this PR, where is it?

Comment (Alan-Jowett) [reply netebpfext/net_ebpf_ext.c:36]: It's generated by encode_program_information.

Problem:
We can't call RPC to encode the type information from the kernel.

Solution:
1) Call RPC from user mode.
2) Store the resulting buffer in header file.
3) Use encoded buffer from driver.

Given that the program information is defined at compile time, we can get away with it.

We still need to solve serialization for the IOCTL path, but this should unblock the verifier changes.

Comment (dthaler) [reply netebpfext/net_ebpf_ext.c:36]: Can you add this explanation as a // comment somewhere in the code?

File: tests/end_to_end/helpers.h (hunk 2 above)
Comment (dthaler) [reply tests/end_to_end/helpers.h:150]: @Alan-Jowett this question was marked resolved but I didn't see a change or a response here so unresolving

Comment (dthaler) [reply tests/end_to_end/helpers.h:150]: Got it, thanks!

**Pull request merged**